File: android/src/main/kotlin/dev/pam/featureflags/FeatureFlagSnapshotModule.kt
package dev.pam.featureflags

import android.content.Context
import dev.pam.nativeapp.modules.ModuleCompletion
import dev.pam.nativeapp.modules.ModuleResultStatus
import dev.pam.nativeapp.modules.NativeModule
import dev.pam.nativeapp.protocol.WireMap
import dev.pam.nativeapp.protocol.WireValue

class FeatureFlagSnapshotModule(
    context: Context,
) : NativeModule {
    private val preferences = context.applicationContext.getSharedPreferences(
        "pam-native-feature-flags",
        Context.MODE_PRIVATE,
    )

    override fun invoke(method: String, payload: ByteArray, completion: ModuleCompletion) {
        runCatching {
            val values = WireMap.decode(payload)
            val namespace = (values["namespace"] as? WireValue.Text)?.value
                ?.takeIf { it.matches(Regex("[A-Za-z0-9._-]{1,128}")) }
                ?: error("Invalid snapshot namespace")
            when (method) {
                "save" -> {
                    val snapshot = (values["snapshot"] as? WireValue.Text)?.value
                        ?: error("Snapshot is required")
                    require(snapshot.toByteArray(Charsets.UTF_8).size <= 1024 * 1024) {
                        "Snapshot exceeds one MiB"
                    }
                    check(preferences.edit().putString(namespace, snapshot).commit()) {
                        "Snapshot commit failed"
                    }
                    WireMap.encode(mapOf("saved" to WireValue.Flag(true)))
                }
                "load" -> WireMap.encode(mapOf(
                    "snapshot" to WireValue.Text(preferences.getString(namespace, "") ?: ""),
                ))
                "delete" -> {
                    check(preferences.edit().remove(namespace).commit()) {
                        "Snapshot delete failed"
                    }
                    WireMap.encode(mapOf("deleted" to WireValue.Flag(true)))
                }
                else -> error("Unknown method: $method")
            }
        }.onSuccess { result ->
            completion.complete(ModuleResultStatus.SUCCESS, result)
        }.onFailure { error ->
            completion.complete(
                ModuleResultStatus.FAILURE,
                (error.message ?: "Feature flag snapshot failed").toByteArray(Charsets.UTF_8),
            )
        }
    }
}

File: ios/Sources/FeatureFlagSnapshotModule.swift
import Foundation
import PamNative

public final class FeatureFlagSnapshotModule: NativeModule, @unchecked Sendable {
    private let defaults: UserDefaults
    private let maxBytes = 1024 * 1024

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    public func invoke(method: String, payload: Data, completion: @escaping ModuleCompletion) {
        do {
            let values = try WireMap.decode(payload)
            guard case let .text(namespace)? = values["namespace"],
                  namespace.range(of: "^[A-Za-z0-9._-]{1,128}$", options: .regularExpression) != nil
            else {
                throw SnapshotError.invalidNamespace
            }
            switch method {
            case "save":
                guard case let .text(snapshot)? = values["snapshot"],
                      snapshot.utf8.count <= maxBytes else {
                    throw SnapshotError.invalidSnapshot
                }
                defaults.set(snapshot, forKey: key(namespace))
                completion(.success, try WireMap.encode(["saved": .flag(true)]))
            case "load":
                let snapshot = defaults.string(forKey: key(namespace)) ?? ""
                completion(.success, try WireMap.encode(["snapshot": .text(snapshot)]))
            case "delete":
                defaults.removeObject(forKey: key(namespace))
                completion(.success, try WireMap.encode(["deleted": .flag(true)]))
            default:
                completion(.failure, Data("Unknown method: \(method)".utf8))
            }
        } catch {
            completion(.failure, Data(String(describing: error).utf8))
        }
    }

    private func key(_ namespace: String) -> String {
        "dev.pam.feature-flags.\(namespace)"
    }
}

private enum SnapshotError: Error {
    case invalidNamespace
    case invalidSnapshot
}
